### PR TITLE
Keep order of tasks in the README notebook

### LINKED ISSUE #285: Keep order of tasks in the README notebook
Right now the pydoc snippets are inserted randomly for each step and random across steps.

### PR DESCRIPTION
**Pull request by larsgeorge-db**
Resolves #285

### CODE REVIEW

**Comment by larsgeorge-db** [comment]
Works

image

File: src/databricks/labs/ucx/framework/tasks.py [hunk 1]
```diff
@@ -63,6 +64,7 @@ def wrapper(*args, **kwargs):
             notebook=notebook,
             dashboard=dashboard,
         )
+        _TASK_ORDER.append(func.__name__)
```
Comment (nfx) [inline src/databricks/labs/ucx/framework/tasks.py:67]: Why do we need to introduce a separate data structure, that has more potential for getting out of sync due to a bug?

Why not a field in the Task class, that holds the definition order (and sorting the tasks before the readme generation) or a topological sort based on depends_on field?

File: src/databricks/labs/ucx/install.py [hunk 2]
```diff
@@ -248,21 +248,29 @@ def _create_jobs(self):
...
     def _create_readme(self):
         md = [
             "# UCX - The Unity Catalog Migration Assistant",
             f'To troubleshoot, see [debug notebook]({self._notebook_link(f"{self._install_folder}/DEBUG.py")}).\n',
             "Here are the URL and descriptions of jobs that trigger's various stages of migration.",
             "All jobs are defined with necessary cluster configurations and DBR versions.",
         ]
-        for step_name, job_id in self._deployed_steps.items():
+        for step_name in self._step_list():
+            job_id = self._deployed_steps[step_name]
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:265]: And what will happen if step is not deployed?

File: tests/unit/test_install.py [hunk 3]
```diff
@@ -220,3 +220,55 @@ def test_choices_happy(mocker):
...
     res = install._choice("foo", ["a", "b"])
     assert "b" == res
+
+
+def test_step_list(mocker):
+    ws = mocker.Mock()
+    install = Installer(ws)
+    from databricks.labs.ucx.framework.tasks import _TASKS, Task
+
+    _TASKS.clear()
```
Comment (nfx) [inline tests/unit/test_install.py:230]: Unit tests may be run from multiple threads in a random order, so this one is not deterministic.

**Review by nfx — CHANGES_REQUESTED** [review]

**Comment by nfx** [comment]
Also change the PR title to be changelog-friendly

File: src/databricks/labs/ucx/framework/tasks.py (hunk 1 above)
Comment (larsgeorge-db) [reply src/databricks/labs/ucx/framework/tasks.py:67]: The latter will not help with the steps ordering, only task ordering. I am open for suggestions on how to do this better.

File: src/databricks/labs/ucx/install.py (hunk 2 above)
Comment (larsgeorge-db) [reply src/databricks/labs/ucx/install.py:265]: Understood, will address asap.

File: tests/unit/test_install.py (hunk 3 above)
Comment (larsgeorge-db) [reply tests/unit/test_install.py:230]: Darn, I need to find a way to make this testable... in general, using protected fields seems to be complained about but some linters.

File: src/databricks/labs/ucx/framework/tasks.py (hunk 1 above)
Comment (larsgeorge-db) [reply src/databricks/labs/ucx/framework/tasks.py:67]: How about an extra field that holds a generated ID, which is starting at 1 and gets increased for every task added. Then we can sort on that field? @nfx

Comment (nfx) [reply src/databricks/labs/ucx/framework/tasks.py:67]: > How about an extra field that holds a generated ID, which is starting at 1 and gets increased for every task added. Then we can sort on that field? 

that's exactly what I'm asking for :) but start with 0, to be consistent with array indexing

File: src/databricks/labs/ucx/framework/tasks.py [hunk 4]
```diff
@@ -53,7 +55,11 @@ def wrapper(*args, **kwargs):
             msg = f"Task {func.__name__} must have documentation"
             raise SyntaxError(msg)
 
+        global _TASK_ID_GENERATOR
```
Comment (larsgeorge-db) [inline src/databricks/labs/ucx/framework/tasks.py:58]: @nfx The linter does not like this:
```
src/databricks/labs/ucx/framework/tasks.py:58:9: PLW0603 Using the global statement to update `_TASK_ID_GENERATOR` is discouraged
```
Is there a better way?

Comment (nfx) [reply src/databricks/labs/ucx/framework/tasks.py:58]: You can len(_TASKS)

Comment (nfx) [reply src/databricks/labs/ucx/framework/tasks.py:58]: no additional counter to track, right?

File: src/databricks/labs/ucx/install.py [hunk 5]
```diff
@@ -248,21 +248,35 @@ def _create_jobs(self):
         self._create_readme()
         self._create_debug(remote_wheel)
 
+    @staticmethod
+    def _sort_task_list() -> list[Task]:
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:252]:
```suggestion
    def _sorted_tasks() -> list[Task]:
```

`_sort` implies no retun

File: src/databricks/labs/ucx/install.py [hunk 6]
```diff
@@ -248,21 +248,35 @@ def _create_jobs(self):
...
+        return sorted(_TASKS.values(), key=lambda x: x.task_id)
+
+    @classmethod
+    def _step_list(cls) -> list[str]:
+        step_list = []
+        [step_list.append(task.workflow) for task in cls._sort_task_list() if task.workflow not in step_list]
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:258]: can you rewrite it as a for-loop, as you're not assigning a result of list comprehension to a value

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/install.py (hunk 5 above)
Comment (larsgeorge-db) [reply src/databricks/labs/ucx/install.py:252]: Does it? Seems arbitrary.

**Review by nfx — APPROVED** [review]
LGTM for implementation, but fix the tests

**Pull request merged**